[New Architecture - Android] Native Module Uncaught Exception doesn't crash the app

### Minimal reproducible example

https://github.com/krystofwoldrich/expo-release-native-module-crash

### Steps to reproduce

1. Build the Android Release version
2. Click `Crash` button
3. Application have not crash
4. Logcat logs show `ReactNoCrashSoftException`

It only happens on new architecture (with legacy arch the application crashes as expected.)

Suspect https://github.com/expo/expo/blob/e8cf02a7d3794a448e862edc038589fec1a7911d/packages/expo/android/src/rn78/main/expo/modules/ExpoReactHostFactory.kt#L70

ExpoReactHostFactory.handleInstanceException swallows the exception without re-throwing it. I would expect the application to crash so error monitoring tools and users are aware of the exceptions.

````
Unhandled SoftException
com.facebook.react.bridge.ReactNoCrashSoftException: raiseSoftException(getOrCreateDestroyTask()): handleHostException(message = "Crashed in native module")
	at com.facebook.react.runtime.ReactHostImpl.raiseSoftException(ReactHostImpl.java:1025)
	at com.facebook.react.runtime.ReactHostImpl.getOrCreateDestroyTask(ReactHostImpl.java:1625)
	at com.facebook.react.runtime.ReactHostImpl.lambda$destroy$7(ReactHostImpl.java:615)
	at com.facebook.react.runtime.ReactHostImpl.$r8$lambda$uso21_D6dCZdcf-JomVD56kdG4c(Unknown Source:0)
	at com.facebook.react.runtime.ReactHostImpl$$ExternalSyntheticLambda35.call(D8$$SyntheticClass:0)
	at com.facebook.react.runtime.internal.bolts.Task$2.run(Task.java:240)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1156)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:651)
	at java.lang.Thread.run(Thread.java:1119)
Caused by: java.lang.RuntimeException: Crashed in native module
	at com.krystofwoldrich.Repro.MainApplication$reactNativeHost$1$getPackages$1$createNativeModules$1.crash(MainApplication.kt:41)
	at com.facebook.jni.NativeRunnable.run(Native Method)
	at android.os.Handler.handleCallback(Handler.java:995)
	at android.os.Handler.dispatchMessage(Handler.java:103)
	at com.facebook.react.bridge.queue.MessageQueueThreadHandler.dispatchMessage(MessageQueueThreadHandler.kt:20)
	at android.os.Looper.loopOnce(Looper.java:248)
	at android.os.Looper.loop(Looper.java:338)
	at com.facebook.react.bridge.queue.MessageQueueThreadImpl$Companion.startNewBackgroundThread$lambda$1(MessageQueueThreadImpl.kt:175)
	at com.facebook.react.bridge.queue.MessageQueueThreadImpl$Companion.$r8$lambda$ldnZnqelhYFctGaUKkOKYj5rxo4(Unknown Source:0)
	at com.facebook.react.bridge.queue.MessageQueueThreadImpl$Companion$$ExternalSyntheticLambda0.run(D8$$SyntheticClass:0)
````

### Environment

```text
expo-env-info 1.3.3 environment info:
    System:
      OS: macOS 15.5
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 22.15.0 - ~/.volta/tools/image/node/22.15.0/bin/node
      Yarn: 3.6.4 - ~/.volta/tools/image/yarn/3.6.4/bin/yarn
      npm: 10.9.2 - ~/.volta/tools/image/node/22.15.0/bin/npm
      Watchman: 2025.04.28.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.16.2 - /Users/krystof/.rbenv/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 24.4, iOS 18.4, macOS 15.4, tvOS 18.4, visionOS 2.4, watchOS 11.4
    IDEs:
      Android Studio: 2024.3 AI-243.24978.46.2431.13363775
      Xcode: 16.3/16E140 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~53.0.9 => 53.0.9
      expo-router: ~5.0.6 => 5.0.7
      react: 19.0.0 => 19.0.0
      react-dom: 19.0.0 => 19.0.0
      react-native: 0.79.2 => 0.79.2
      react-native-web: ~0.20.0 => 0.20.0
    Expo Workflow: bare
```

### Expo Doctor Diagnostics

```text
14/15 checks passed. 1 checks failed. Possible issues detected:
Use the --verbose flag to see more details about passed checks.

✖ Check for app config fields that may not be synced in a non-CNG project
This project contains native project folders but also has native configuration properties in app.json, indicating it is configured to use Prebuild. When the android/ios folders are present, if you don't run prebuild in your build pipeline, the following properties will not be synced: orientation, icon, scheme, userInterfaceStyle, ios, android, plugins, androidStatusBar.


1 check failed, indicating possible issues with the project.
```

--- a/packages/expo/CHANGELOG.md
+++ b/packages/expo/CHANGELOG.md
@@ -9,6 +9,7 @@
 ### 🐛 Bug fixes
 
 - [iOS] Call `createRootViewController` from the `ExpoReactNativeFactoryDelegate`. ([#36787](https://github.com/expo/expo/pull/36787) by [@alanjhughes](https://github.com/alanjhughes))
+- [Android] Re-throw error in `handleInstanceException` if nothing is registered to handle the error.
 
 ### 💡 Others
 

--- a/packages/expo/CHANGELOG.md
+++ b/packages/expo/CHANGELOG.md
@@ -9,7 +9,7 @@
 ### 🐛 Bug fixes
 
 - [iOS] Call `createRootViewController` from the `ExpoReactNativeFactoryDelegate`. ([#36787](https://github.com/expo/expo/pull/36787) by [@alanjhughes](https://github.com/alanjhughes))
-- [Android] Re-throw error in `handleInstanceException` if nothing is registered to handle the error.
+- [Android] Re-throw error in `handleInstanceException` if nothing is registered to handle it. ([#37021](https://github.com/expo/expo/pull/37021) by [@alanjhughes](https://github.com/alanjhughes))
 
 ### 💡 Others
 

--- a/packages/expo/android/src/rn78/main/expo/modules/ExpoReactHostFactory.kt
+++ b/packages/expo/android/src/rn78/main/expo/modules/ExpoReactHostFactory.kt
@@ -68,8 +68,12 @@ object ExpoReactHostFactory {
       get() = reactNativeHostWrapper.packages
 
     override fun handleInstanceException(error: Exception) {
+      val handlers = reactNativeHostWrapper.reactNativeHostHandlers
+      if (handlers.isEmpty()) {
+        throw error
+      }
       val useDeveloperSupport = reactNativeHostWrapper.useDeveloperSupport
-      reactNativeHostWrapper.reactNativeHostHandlers.forEach { handler ->
+      handlers.forEach { handler ->
         handler.onReactInstanceException(useDeveloperSupport, error)
       }
     }

